wallet: Replace CWalletTx::mapValue and vOrderForm with explicit class members

achow101 opened the pull request:
`mapValue` and `vOrderForm` are opaque data structures that contain transaction metadata. It is hard to determine what actual data each field contains, and they can ostensibly be misused where metadata is added in the future without developers realizing that such metadata exists.

It's much clearer to have all of that metadata live in their own explicit member variables within `CWalletTx`. This PR implements that change.

Since the serialization format of `CWalletTx` depends on `mapValue` and `vOrderForm`, the serialization remains unchanged, so when serializing these new members, they need to be shoved/extracted from a temporary `mapValue` or `vOrderForm`.

This does end up breaking forwards compatibility as unknown fields in `mapValue` and `vOrderForm` are stripped out if the record is rewritten. However, I don't expect that we would continue to use these fields for future metadata, so I think that risk is low.

achow101 committed:
wallet: Pass replaces_txid to CommitTransaction outside of mapValue

Instead of updating mapValue with the "replaces_txid" value and passing
the updated mapValue to CommitTransaction, pass the replaces_txid
directly to CommitTransaction which updates mapValue as necessary.

This is prepration for removing mapValue.

achow101 committed:
wallet: Pass comment and comment_to to CommitTransaction

Instead of passing these by setting them in mapValue, pass them directly
to CommitTransaction.

This is preparation for removing mapValue.

achow101 committed:
wallet: Drop mapValue from CommitTransaction

The values previously passed in mapValue are now parameters to
CommitTransaction so there is no need for mapValue to be passed.

achow101 committed:
wallet: Make CWalletTx "comment" and "to" member variables

Instead of storing "comment" and "to" inside of mapValue, store these
expliclty as members of CWalletTx.

achow101 committed:
wallet: Make CWalletTx "replaces_txid" and "replaced_by_txid" member variables

Instead of storing "replaces_txid" and "replaced_by_txid" as strings inside of
mapValue, store these expliclty as members of CWalletTx.

achow101 committed:
wallet: Drop mapValue from CWalletTx

It doesn't make sense to be storing relevant metadata variables inside
of a string map in CWalletTx. All of the fields have been pulled out
into separate members, so there is no need for mapValue to stick around.

achow101 committed:
wallet: Drop vOrderForm from CommitTransaction

This parameter is only used to pass in the "Messages" from the GUI.
Instead of making it opaque by putting those into vOrderForm, use a
specific dedicated parameter for providing the messages.

achow101 committed:
wallet: Replace CWalletTx's vOrderForm with specific fields

vOrderForm contained 2 kinds of strings: BIP 21 messages, and BIP 70
Payment Requests. Instead of having both inside of a single vOrderForm
field that is opaque, split them into separate std::vector<std::string>
to contain this metadata.

achow101 committed:
wallet: Make CWalletTx "from" and "message" member variables

Instead of storing "from" and "message" inside of mapValue, store these
explicitly as members of CWalletTx.

achow101 commented:
> Is forward compatibility a concern here?

I don't really think so. The last time these were modified was ages ago, I don't foresee any need to continue to use `mapValue` or `vOrderForm` for new fields. Additionally, once a transaction is confirmed, it's unlikely that the record would ever be modified again.

It would be straightforward to preserve `mapValue` and `vOrderForm` for any unknown values, but I think it's unnecessary.

The primary motivation for this change is actually that I am working on a completely new storage mechanism for `CWalletTx` that utilizes SQLite as a SQL database, and I think the best way for that is to have each relevant field be its own column in a new "transactions" table. 

> Would be good to update PR the description to say what happens to unknown fields if a transaction is deserialized and reserialized.

Currently they would be stripped out.

ryanofsky commented:
I guess I'm wondering a little more about forward compatibility. In general, it seems like it could be a useful thing for future wallet software to be able to add metadata to transactions that older software will preserve and ignore. If the deserialization code is changed to now discard unrecognized mapvalue fields instead of preserving them, are there different places new transaction metadata could be stored safely in the future? Or is mapvalue's extensibiilty just less useful than it might seem, and if we added new transaction fields maybe we'd always want to forbid opening the new wallets with old software anyway? Alternately I'd wonder if `CWalletTx::Unserialize` should raise an error if an unrecognized field is present, so it does not appear to deserialize successfully when it might discard information. Preserving unknown fields does seems like the most conservative option here, but you know a lot more about this, so feel free to go with whatever approach make the most sense.

achow101 commented:
> Or is mapvalue's extensibiilty just less useful than it might seem

This, generally. What we usually do for upgrading metadata is to put the new metadata at the end of the record. Deserialization always allows for there to be extra data, so it's always backwards compatible to stick things onto the end. The only compatibility issue is if the record is rewritten during a downgrade followed by a re-upgrade. But we also have the wallet flags that we use for downgrade protection if that is necessary.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32763/commits/9a8fd16e67755e336de683c0158d4fdd2b4ddfe8.
`value_map` and `order_form` (and therefore `WalletOrderForm` and `WalletValueMap`) lack clarity, making replacing with well-defined fields a much better approach.

laanwj commented:
Concept ACK

> I don't think a new field has been added to mapValue or vOrderForm in about a decade too, and I don't think anyone is planning to add any new fields any time soon.

i agree here. And it's a real messy, non-type-safe way to handle metadata. 

We have also never offered setting these directly on the RPC interface, so there's negligible chance of unknown ones existing anywhere. 

When we do need to add metadata, it's better to add and serialize it as actual fields of the transaction structure, not stash it into a map like this.

ajtowns commented on src/wallet/transaction.h:
```diff
@@ -324,27 +293,19 @@ class CWalletTx
             else if (key == "comment") m_comment = value;
             else if (key == "to") m_comment_to = value;
             else if (key == "replaces_txid") m_replaces_txid = Txid::FromHex(value);
             else if (key == "replaced_by_txid") m_replaced_by_txid = Txid::FromHex(value);
```
I think it would be better if there was some sort of `else throw("Invalid mapValue key in wallet file");` detection here that your wallet file had extraneous data that isn't being parsed and will be lost? If there's no software out there using random stuff in mapValue, then the else will never trigger; if there is, it might save them from losing data they consider important. Easier to write a migration tool if their data is preserved and the upgrade breaks, than a recovery tool if the upgrade works but the data is lost.

achow101 replied:
Done

ajtowns reviewed (commented):
Approach ACK c79c1d4845399b1f0a945addb0287bec5b07ec46

achow101 committed:
wallet: Throw if unknown entry is found in mapValue

We know what entries are expected to be found in mapValue. If we
encounter one that is not recognized, then we should throw and refuse to
load the wallet in order to avoid losing user data once mapValue is no
longer used.

ajtowns commented on src/wallet/transaction.h:
```diff
@@ -194,6 +194,11 @@ class CMerkleTx
 public:
+    // "from" and "message" are obsolete fields that could be set in
+    // the UI prior to 2011 (removed in commit 4d9b223)
+    // These fields are kept to avoid losing metadata.
```
They aren't serialized or deserialized though?

achow101 replied:
Huh, missed that somehow. Fixed

ajtowns commented on src/wallet/transaction.h:
```diff
@@ -284,14 +267,22 @@ class CWalletTx
+        if (nTimeSmart) string_values["timesmart"] = strprintf("%u", nTimeSmart);
+
+        std::vector<std::pair<std::string, std::string>> msgs_reqs;
+        msgs_reqs.resize(m_messages.size() + m_payment_requests.size());
```
reserve not resize?

achow101 replied:
Fixed

Eunovo commented on src/wallet/wallet.cpp:
```diff
@@ -764,6 +764,8 @@ void CWallet::SyncMetaData(std::pair<TxSpends::iterator, TxSpends::iterator> ran
         if (copyFrom == copyTo) continue;
         assert(copyFrom && "Oldest wallet transaction in range assumed to have been found.");
         if (!copyFrom->IsEquivalentTo(*copyTo)) continue;
+        copyTo->m_comment = copyFrom->m_comment;
```
https://github.com/bitcoin/bitcoin/pull/32763/commits/e82ad0148442219c9d9778bd8f6d84b6bc89f7eb:

I noticed that you didn't copy `m_from` and `m_message`, is this intentional?

achow101 replied:
Fixed

Eunovo reviewed (commented):
Concept ACK https://github.com/bitcoin/bitcoin/pull/32763/commits/69088787f545d351b74e9da12f9c8f26aff4715d:

Speaking as someone relatively new to the project, `mapValue` and `vOrderForm` were "black boxes" to me.

w0xlt commented on src/wallet/transaction.h:
```diff
@@ -337,6 +337,8 @@ class CWalletTx
             else if (key == "comment") m_comment = value;
             else if (key == "to") m_comment_to = value;
+            else if (key == "replaces_txid") m_replaces_txid = Txid::FromHex(value);
+            else if (key == "replaced_by_txid") m_replaced_by_txid = Txid::FromHex(value);
```
Should this validate the parsed txid? 

`Txid::FromHex(value)` returns `std::nullopt` for malformed input, so assigning it directly makes an invalid stored `replaces_txid` / `replaced_by_txid` indistinguishable from a missing key. If the wallet is later rewritten, that metadata is silently lost.

```diff
diff --git a/src/wallet/transaction.h b/src/wallet/transaction.h
index 3697da8e35..ee3a5a2001 100644
--- a/src/wallet/transaction.h
+++ b/src/wallet/transaction.h
@@ -322,9 +322,15 @@ public:
             else if (key == "message") m_message = value;
             else if (key == "comment") m_comment = value;
             else if (key == "to") m_comment_to = value;
-            else if (key == "replaces_txid") m_replaces_txid = Txid::FromHex(value);
-            else if (key == "replaced_by_txid") m_replaced_by_txid = Txid::FromHex(value);
-            else {
+            else if (key == "replaces_txid") {
+                const auto txid{Txid::FromHex(value)};
+                if (!txid) throw std::runtime_error("Invalid replaces_txid value in CWalletTx string value map");
+                m_replaces_txid = *txid;
+            } else if (key == "replaced_by_txid") {
+                const auto txid{Txid::FromHex(value)};
+                if (!txid) throw std::runtime_error("Invalid replaced_by_txid value in CWalletTx string value map");
+                m_replaced_by_txid = *txid;
+            } else {
                 throw std::runtime_error("Unexpected value in CWalletTx strings value map");
             }
         }

```

Unit test (as an example):

<details>
  <summary>diff</summary>

```diff
diff --git a/src/wallet/test/wallet_transaction_tests.cpp b/src/wallet/test/wallet_transaction_tests.cpp
index 38be52f41f..da3e4d1d3d 100644
--- a/src/wallet/test/wallet_transaction_tests.cpp
+++ b/src/wallet/test/wallet_transaction_tests.cpp
@@ -4,6 +4,7 @@
 
 #include <wallet/transaction.h>
 
+#include <streams.h>
 #include <test/util/common.h>
 #include <wallet/test/wallet_test_fixture.h>
 
@@ -23,5 +24,33 @@ BOOST_AUTO_TEST_CASE(roundtrip)
     }
 }
 
+BOOST_AUTO_TEST_CASE(malformed_replacement_txids_throw)
+{
+    auto check_invalid_txid = [](const std::string& key) {
+        const CTransactionRef tx{MakeTransactionRef(CMutableTransaction{})};
+        const TxState state{TxStateInactive{}};
+        const std::map<std::string, std::string> string_values{{key, "not-a-txid"}};
+
+        DataStream stream{};
+        stream << TX_WITH_WITNESS(tx)
+               << TxStateSerializedBlockHash(state)
+               << std::vector<uint8_t>{}
+               << TxStateSerializedIndex(state)
+               << std::vector<uint8_t>{}
+               << string_values
+               << std::vector<std::pair<std::string, std::string>>{}
+               << uint32_t{0}
+               << 0U
+               << false
+               << false;
+
+        CWalletTx wtx{nullptr, TxStateInactive{}};
+        BOOST_CHECK_THROW(stream >> wtx, std::runtime_error);
+    };
+
+    check_invalid_txid("replaces_txid");
+    check_invalid_txid("replaced_by_txid");
+}
+
 BOOST_AUTO_TEST_SUITE_END()
 } // namespace wallet
```
</details>

achow101 replied:
I don't think that is necessary, if it is corrupted, then it is as good as lost. This information is not something that can be recovered.

w0xlt commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -62,8 +62,10 @@ static void WalletTxToJSON(const CWallet& wallet, const CWalletTx& wtx, UniValue
-        entry.pushKV(item.first, item.second);
+    if (wtx.m_comment) entry.pushKV("comment", *wtx.m_comment);
+    if (wtx.m_comment_to) entry.pushKV("to", *wtx.m_comment_to);
+    if (wtx.m_replaces_txid) entry.pushKV("replaces_txid", wtx.m_replaces_txid->ToString());
```
`WalletTxToJSON()` no longer returns the legacy `from` and `message` metadata fields. 

Assuming this behavior change is not intentional, the patch below restores the previous RPC JSON behavior. The fields are still preserved and displayed in the GUI, so omitting them here seems accidental.

```diff
diff --git a/src/wallet/rpc/transactions.cpp b/src/wallet/rpc/transactions.cpp
index 480a346f3c..452f350e8c 100644
--- a/src/wallet/rpc/transactions.cpp
+++ b/src/wallet/rpc/transactions.cpp
@@ -64,6 +64,8 @@ static void WalletTxToJSON(const CWallet& wallet, const CWalletTx& wtx, UniValue
 
     if (wtx.m_comment) entry.pushKV("comment", *wtx.m_comment);
     if (wtx.m_comment_to) entry.pushKV("to", *wtx.m_comment_to);
+    if (wtx.m_from) entry.pushKV("from", *wtx.m_from);
+    if (wtx.m_message) entry.pushKV("message", *wtx.m_message);
     if (wtx.m_replaces_txid) entry.pushKV("replaces_txid", wtx.m_replaces_txid->ToString());
     if (wtx.m_replaced_by_txid) entry.pushKV("replaced_by_txid", wtx.m_replaced_by_txid->ToString());
 }
````

achow101 replied:
These fields are not documented and so deprecated that the fact they would be output but not documented that they don't result in any test failures even though they would. I don't think we should be outputting these.

w0xlt replied:
I think this creates an inconsistency between RPC and GUI behavior: `from` and `message` are still preserved and displayed in the GUI, but no longer returned by RPC. This is not a blocker, though.

w0xlt commented on src/wallet/transaction.h:
```diff
@@ -257,8 +242,6 @@ class CWalletTx
     {
-        mapValue.clear();
-        vOrderForm.clear();
         nTimeReceived = 0;
```
Before this PR, `Init()` cleared the metadata containers:

* `mapValue.clear();`
* `vOrderForm.clear();`

After replacing those containers with explicit fields, `Init()` no longer clears the new fields. If `Unserialize()` is called on a reused `CWalletTx`, this can leave stale metadata from the previous value attached to the newly deserialized transaction.

Suggested change:

```diff
diff --git a/src/wallet/transaction.h b/src/wallet/transaction.h
index 3697da8e35..5456f4ba4a 100644
--- a/src/wallet/transaction.h
+++ b/src/wallet/transaction.h
@@ -242,6 +242,14 @@ public:
 
     void Init()
     {
+        m_from.reset();
+        m_message.reset();
+        m_comment.reset();
+        m_comment_to.reset();
+        m_replaces_txid.reset();
+        m_replaced_by_txid.reset();
+        m_messages.clear();
+        m_payment_requests.clear();
         nTimeReceived = 0;
         nTimeSmart = 0;
         fChangeCached = false;
```

Test:

<details>
  <summary>diff</summary>

```diff
diff --git a/src/wallet/test/wallet_transaction_tests.cpp b/src/wallet/test/wallet_transaction_tests.cpp
index 38be52f41f..bb2641120d 100644
--- a/src/wallet/test/wallet_transaction_tests.cpp
+++ b/src/wallet/test/wallet_transaction_tests.cpp
@@ -4,6 +4,7 @@
 
 #include <wallet/transaction.h>
 
+#include <streams.h>
 #include <test/util/common.h>
 #include <wallet/test/wallet_test_fixture.h>
 
@@ -23,5 +24,33 @@ BOOST_AUTO_TEST_CASE(roundtrip)
     }
 }
 
+BOOST_AUTO_TEST_CASE(unserialize_clears_metadata)
+{
+    const auto tx{MakeTransactionRef(CMutableTransaction{})};
+    CWalletTx source{tx, TxStateInactive{}};
+    DataStream stream{};
+    stream << source;
+
+    CWalletTx target{tx, TxStateInactive{}};
+    target.m_from = "from";
+    target.m_message = "message";
+    target.m_comment = "comment";
+    target.m_comment_to = "to";
+    target.m_replaces_txid = Txid::FromUint256(uint256::ONE);
+    target.m_replaced_by_txid = Txid::FromUint256(uint256::ONE);
+    target.m_messages = {"message"};
+    target.m_payment_requests = {"payment_request"};
+
+    stream >> target;
+    BOOST_CHECK(!target.m_from);
+    BOOST_CHECK(!target.m_message);
+    BOOST_CHECK(!target.m_comment);
+    BOOST_CHECK(!target.m_comment_to);
+    BOOST_CHECK(!target.m_replaces_txid);
+    BOOST_CHECK(!target.m_replaced_by_txid);
+    BOOST_CHECK(target.m_messages.empty());
+    BOOST_CHECK(target.m_payment_requests.empty());
+}
+
 BOOST_AUTO_TEST_SUITE_END()
 } // namespace wallet
```
</details>

achow101 replied:
`Init()` is really a vestigial function. `CWalletTx` should never be reused and doing so is a programming error. #35501 moves us more strongly in that direction.

ajtowns commented on src/wallet/transaction.h:
```diff
@@ -194,33 +191,21 @@ class CMerkleTx
+    std::vector<std::string> m_messages;
+    // BIP 70 Payment Request (deprecated, field kept to preserve metadata from old wallets)
+    std::vector<std::string> m_payment_requests;
+    unsigned int fTimeReceivedIsTxTime;
```
`fTimeReceivedIsTxTime` is unused; bad rebase vs #32768 somewhere?

achow101 replied:
Oops fixed

ajtowns commented on src/wallet/wallet.cpp:
```diff
@@ -2347,7 +2348,16 @@ void CWallet::CommitTransaction(
+            }
+        }
+        if (!payment_requests.empty()) {
+            for (const std::string& msg : messages) {
```
Should be iterating over `payment_requests`. But this code is removed in the next commit, so nbd.

achow101 replied:
Fixed

ajtowns reviewed (commented):
ACK fbee36b8392720b0d2da7b49a610e0264c7e79b1 ; nothing blocking here afaics

w0xlt reviewed (commented):
ACK fed3cf6f0ed40ca9720d8f98b26f1b08afc93275 with above nits/caveats.

Eunovo commented on src/wallet/transaction.h:
```diff
@@ -339,10 +341,11 @@ class CWalletTx
             else if (key == "replaces_txid") m_replaces_txid = Txid::FromHex(value);
             else if (key == "replaced_by_txid") m_replaced_by_txid = Txid::FromHex(value);
+            else {
+                throw std::runtime_error("Unexpected value in CWalletTx strings value map");
```
https://github.com/bitcoin/bitcoin/pull/32763/commits/cb99864c91c9def07f75872df533a5bd7a3c649b:

Might be a good idea to add a comment here stating that this doesn't affect backwards compatibility because new metadata fields should only be appended to the WalletTx serialised data instead of using the map.

achow101 replied:
If I need to retouch.

Eunovo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/32763/commits/fed3cf6f0ed40ca9720d8f98b26f1b08afc93275

ajtowns commented:
reACK fed3cf6f0ed40ca9720d8f98b26f1b08afc93275
